### PR TITLE
nspawn: don't hide --bind=/tmp/* mounts

### LINKED ISSUE #4789: systemd-nspawn fails to --bind directories
### Submission type

  - [X] Bug report
  - [ ] Request for enhancement (RFE)

*NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!*

### systemd version the issue has been seen with

> 232

*NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!*

### Used distribution

> Arch Linux

### In case of bug report: Expected behaviour you didn't see

> systemd-nspawn fails to --bind=/tmp/.X11-unix into the container

### In case of bug report: Unexpected behaviour you saw

> none

### In case of bug report: Steps to reproduce the problem

> btrfs subvolume snapshot / /.nspawn ; systemd-nspawn -D /.nspawn --bind=/tmp/.X11-unix bash
> In 231 this shows /tmp/.X11-unix in the container, in 232 /tmp is empty.
> I've attached the 232 output with LOG_LEVEL=debug

[nspawn.txt](https://github.com/systemd/systemd/files/625273/nspawn.txt)

### PR DESCRIPTION
**Pull request by evverx**
Fixes #4789

### CODE REVIEW

**Comment by evverx** [comment]
```
[systemd-pr-build] $ /bin/sh -xe /tmp/hudson2438376146086078148.sh
+ /home/systemd/jenkins-glue.sh
Traceback (most recent call last):
  File "./slave-control.py", line 193, in <module>
    main()
  File "./slave-control.py", line 131, in main
    data = json.loads(json_data)
  File "/usr/lib64/python2.7/json/__init__.py", line 338, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python2.7/json/decoder.py", line 365, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python2.7/json/decoder.py", line 383, in raw_decode
    raise ValueError("No JSON object could be decoded")
ValueError: No JSON object could be decoded
```
@zonque , please take a look.
[`json_data = duffy_cmd("/Node/get", params)`](https://github.com/systemd/systemd-centos-ci/blob/99eb550beffd39618e1a63295edecedf2ffcf7dc/slave-control.py#L130) is not working

File: src/nspawn/nspawn-mount.c [hunk 1]
```diff
@@ -550,7 +550,7 @@ int mount_all(const char *dest,
                 { NULL,                  "/proc/sys",           NULL,    NULL,        MS_BIND|MS_RDONLY|MS_NOSUID|MS_NOEXEC|MS_NODEV|MS_REMOUNT, MOUNT_FATAL|MOUNT_IN_USERNS|MOUNT_APPLY_APIVFS_RO },                          /* ... then, make it r/o */
                 { "/proc/sysrq-trigger", "/proc/sysrq-trigger", NULL,    NULL,        MS_BIND,                                                               MOUNT_IN_USERNS|MOUNT_APPLY_APIVFS_RO },                          /* Bind mount first ...*/
                 { NULL,                  "/proc/sysrq-trigger", NULL,    NULL,        MS_BIND|MS_RDONLY|MS_NOSUID|MS_NOEXEC|MS_NODEV|MS_REMOUNT,             MOUNT_IN_USERNS|MOUNT_APPLY_APIVFS_RO },                          /* ... then, make it r/o */
-                { "tmpfs",               "/tmp",                "tmpfs", "mode=1777", MS_STRICTATIME,                                            MOUNT_FATAL|MOUNT_IN_USERNS },
+                { "tmpfs",               "/tmp",                "tmpfs", "mode=1777", MS_STRICTATIME,                                            MOUNT_FATAL },
```
Comment (evverx) [inline src/nspawn/nspawn-mount.c]: I should move this line to the "outer child"-part. Will fix

**Comment by martinpitt** [comment]
Fedora test is broken, rest is okay.

**Comment by haraldh** [comment]
@systemd-centos-ci 
retest this please

**Comment by poettering** [comment]
All looks good. merging.

@evverx i wonder though whether we can find a fix for the problem you tried to fix with 8492849? I mean, there was a good reason to make that change after all.

I wonder if there's a nice way to detect "foreign" mounts, so that PID 1 in the container could exclude them from umount.target Conflicts= lines. Or alternatively, we could add some logic that makes /tmp a user mount only when it is safe. I.e. basicely reuse your suggested logic, but conditionalize it on the whether there are any mounts configured with /tmp as prefix?

**Comment by evverx** [comment]
> I wonder if there's a nice way to detect "foreign" mounts, so that PID 1 in the container could exclude them from umount.target Conflicts= lines

I think the only reliable way is to ask `nspawn` via some communication channel. But that's the `nspawn`-specific way and it doesn't fix something like https://github.com/systemd/systemd/issues/4810

**Comment by poettering** [comment]
> I think the only reliable way is to ask nspawn via some communication channel. But that's the nspawn-specific way and it doesn't fix something like #4810

I wonder if /proc/self/mountinfo in some way let's us know this...

**Comment by evverx** [comment]
Hm, that's a good question. I didn't think about it.

**Pull request merged**